pkg/secret/monitor.go

```diff
@@ -0,0 +1,140 @@
+package secret
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+// ObjectKey represents the unique identifier for a resource, used to access it in the cache.
+type ObjectKey struct {
+	// Namespace is the namespace in which the resource is located.
+	Namespace string
+	// Name denotes metadata.name of a resource being monitored by informer
+	Name string
+}
+
+// singleItemMonitor monitors a single resource using a SharedInformer.
+type singleItemMonitor struct {
+	key      ObjectKey
+	informer cache.SharedInformer
+	lock     sync.Mutex
+	stopped  bool
+	stopCh   chan struct{}
+}
+
+// NewObjectKey creates a new ObjectKey for the given namespace and name.
+func NewObjectKey(namespace, name string) ObjectKey {
+	return ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}
+}
+
+// newSingleItemMonitor creates a new singleItemMonitor for the given key and informer.
+func newSingleItemMonitor(key ObjectKey, informer cache.SharedInformer) *singleItemMonitor {
+	return &singleItemMonitor{
+		key:      key,
+		informer: informer,
+		stopped:  true,
+		stopCh:   make(chan struct{}),
+	}
+}
+
+// HasSynced returns true if the informer's cache has been successfully synced.
+func (i *singleItemMonitor) HasSynced() bool {
+	return i.informer.HasSynced()
+}
+
+// StartInformer starts and runs the informer until the provided context is canceled,
+// or StopInformer() is called. It will block, so call via goroutine.
+func (i *singleItemMonitor) StartInformer(ctx context.Context) {
+	i.lock.Lock()
+
+	if !i.stopped {
+		klog.Warning("informer is already running")
+		i.lock.Unlock()
+		return
+	}
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			klog.V(5).Info("stopping informer due to context cancellation")
+			if !i.StopInformer() {
+				klog.Error("failed to stop informer")
+			}
+		// this case is required to exit from the goroutine
+		// after normal StopInformer() call i.e when stopCh is closed.
+		case <-i.stopCh:
+		}
+	}()
+
+	klog.Info("starting informer")
+	i.stopped = false
+	i.lock.Unlock()
+
+	i.informer.Run(i.stopCh)
```

Comment:
```suggestion
	go i.informer.Run(i.stopCh)
```

to avoid having your start block.

Reply:
It's blocking by design, the comment states this:

> It will block, so call via goroutine.

Reply:
Updated to make it non-blocking